src/lib.rs

```diff
@@ -0,0 +1,560 @@
+
+extern crate rand;
+
+/// A node inside a `SearchTree`.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Node<T: PartialEq> {
+    pub value: T,
+    parent: Option<usize>,
+    left_child: Option<usize>,
+    right_child: Option<usize>,
+    height: usize,
+}
+
+/// A simple AVL tree.
+#[derive(Debug, PartialEq, Default)]
+pub struct SearchTree <T: PartialEq + PartialOrd> {
+    nodes: Vec<Node<T>>,
+    root: Option<usize>,
+}
+
+impl<'a, T> SearchTree<T>
+    where T: PartialEq + PartialOrd + Clone {
+    /// Creates an empty search tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let tree: SearchTree<String> = SearchTree::new();
+    /// assert!(tree.is_empty());
+    /// ```
+    pub fn new() -> SearchTree<T> {
+        SearchTree {
+            nodes: Vec::new(),
+            root: None
+        }
+    }
+
+    /// Test if this tree is empty.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// assert!(tree.is_empty());
+    /// tree.insert(String::from("foo"));
+    /// assert!(!tree.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.root.is_none()
+    }
+
+    /// Test if this contains a given value.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// tree.insert(String::from("hello"));
+    /// tree.insert(String::from("world!"));
+    /// assert!(tree.contains(&String::from("hello")));
+    /// assert!(!tree.contains(&String::from("cruel")));
+    /// ```
+    pub fn contains(&self, value: &T) -> bool {
+        self.get_node(value).is_ok()
+    }
+
+    /// Add the specified value to this tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// tree.insert(String::from("hello"));
+    /// tree.insert(String::from("world!"));
+    /// assert!(tree.contains(&String::from("hello")));
+    /// assert!(tree.contains(&String::from("world!")));
+    /// ```
+    pub fn insert(&mut self, value: T) {
+        if let Ok(node) = self.create_node(value) {
+            self.nodes.push(node);
+            // inserted the root
+            if self.nodes.len() == 1 {
+                self.root = Some(0);
+            }
+            let last_added = self.nodes.len() - 1;
+            self.update_heights(last_added);
+            let mut node = Some(last_added);
+            while let Some(index) = node {
+                if !self.is_balanced(index) {
+                    self.rebalance(index);
+                }
+                node = self.nodes[index].parent;
+            }
+        }
+    }
+
+    fn get_node(&self, value: &T) -> Result<usize, Option<usize>> {
+        let mut current_index = self.root;
+        let mut previous_index = None;
+        if !self.nodes.is_empty() {
```

Comment:
I'm not sure this check is needed. In the case of the tree being empty, `root` would be None, and thus he `while` loop would not be entered, right?

Reply:
Good point!